pkg/proxy/healthcheck/service_health.go

```diff
@@ -52,7 +52,7 @@ type ServiceHealthServer interface {
 	SyncEndpoints(newEndpoints map[types.NamespacedName]int) error
 }
 
-func newServiceHealthServer(hostname string, recorder events.EventRecorder, listener listener, factory httpServerFactory, nodePortAddresses []string) ServiceHealthServer {
+func newServiceHealthServer(hostname string, recorder events.EventRecorder, listener listener, factory httpServerFactory, nodePortAddresses []string, healthzServer ProxierHealthUpdater) ServiceHealthServer {
```

Comment:
I would maybe do:

```
type proxyHealthChecker interface {
  IsHealthy() bool
}

func newServiceHealthServer(..., proxyChecker proxyHealthChecker) ServiceHealthServer {
```

Reply:
We could, [but every proxier implementation is initialized with a `ProxierHealthUpdater`](https://github.com/kubernetes/kubernetes/blob/f286545c48fa04f609cb923ac39b5c7679098054/pkg/proxy/iptables/proxier.go#L237), so we would need to do wire things like:

```go
// ProxierHealthUpdater allows callers to update healthz timestamp only.
type ProxierHealthUpdater interface {
	// QueuedUpdate should be called when the proxier receives a Service or Endpoints
	// event containing information that requires updating service rules.
	QueuedUpdate()

	// Updated should be called when the proxier has successfully updated the service
	// rules to reflect the current state.
	Updated()

	// Run starts the healthz HTTP server and blocks until it exits.
	Run() error

	proxierHealthChecker
}

type proxierHealthChecker interface {
	// IsHealthy returns the proxier's health state, following the same
	// definition the HTTP server defines.
	IsHealthy() bool
}
```

maybe I'm missing something?

Reply:
I just meant an interface to decouple and maybe make tesing easier.

I checked out your PR and made this change:

```diff
diff --git a/pkg/proxy/healthcheck/service_health.go b/pkg/proxy/healthcheck/service_health.go
index 277a127d9dc..44f3d255afe 100644
--- a/pkg/proxy/healthcheck/service_health.go
+++ b/pkg/proxy/healthcheck/service_health.go
@@ -52,7 +52,12 @@ type ServiceHealthServer interface {
 	SyncEndpoints(newEndpoints map[types.NamespacedName]int) error
 }
 
-func newServiceHealthServer(hostname string, recorder events.EventRecorder, listener listener, factory httpServerFactory, nodePortAddresses []string, healthzServer ProxierHealthUpdater) ServiceHealthServer {
+// healthChecker is something that can report whether it is healthy or not.
+type healthChecker interface {
+	IsHealthy() bool
+}
+
+func newServiceHealthServer(hostname string, recorder events.EventRecorder, listener listener, factory httpServerFactory, nodePortAddresses []string, proxyChecker healthChecker) ServiceHealthServer {
 
 	nodeAddresses, err := utilproxy.GetNodeAddresses(nodePortAddresses, utilproxy.RealNetwork{})
 	if err != nil || nodeAddresses.Len() == 0 {
@@ -75,15 +80,15 @@ func newServiceHealthServer(hostname string, recorder events.EventRecorder, list
 		recorder:      recorder,
 		listener:      listener,
 		httpFactory:   factory,
-		healthzServer: healthzServer,
+		proxyChecker:  proxyChecker,
 		services:      map[types.NamespacedName]*hcInstance{},
 		nodeAddresses: nodeAddresses,
 	}
 }
 
 // NewServiceHealthServer allocates a new service healthcheck server manager
-func NewServiceHealthServer(hostname string, recorder events.EventRecorder, nodePortAddresses []string, healthzServer ProxierHealthUpdater) ServiceHealthServer {
-	return newServiceHealthServer(hostname, recorder, stdNetListener{}, stdHTTPServerFactory{}, nodePortAddresses, healthzServer)
+func NewServiceHealthServer(hostname string, recorder events.EventRecorder, nodePortAddresses []string, proxyChecker healthChecker) ServiceHealthServer {
+	return newServiceHealthServer(hostname, recorder, stdNetListener{}, stdHTTPServerFactory{}, nodePortAddresses, proxyChecker)
 }
 
 type server struct {
@@ -94,7 +99,7 @@ type server struct {
 	listener      listener
 	httpFactory   httpServerFactory
 
-	healthzServer ProxierHealthUpdater
+	proxyChecker healthChecker
 
 	lock     sync.RWMutex
 	services map[types.NamespacedName]*hcInstance
@@ -229,7 +234,7 @@ func (h hcHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
 		return
 	}
 	count := svc.endpoints
-	kubeProxyHealthy := h.hcs.healthzServer.IsHealthy()
+	kubeProxyHealthy := h.hcs.proxyChecker.IsHealthy()
 	h.hcs.lock.RUnlock()
 
 	resp.Header().Set("Content-Type", "application/json")
```

That compiles.

Reply:
OK, updated the PR with this